File: app/src/main/java/org/tasks/jobs/RefreshWork.java
package org.tasks.jobs;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.work.WorkerParameters;
import javax.inject.Inject;
import org.tasks.LocalBroadcastManager;
import org.tasks.injection.JobComponent;
import org.tasks.scheduling.RefreshScheduler;

public class RefreshWork extends RepeatingWorker {

  @Inject RefreshScheduler refreshScheduler;
  @Inject LocalBroadcastManager localBroadcastManager;

  public RefreshWork(@NonNull Context context, @NonNull WorkerParameters workerParams) {
    super(context, workerParams);
  }

  @NonNull
  @Override
  public Result run() {
    localBroadcastManager.broadcastRefresh();
    return Result.success();
  }

  @Override
  protected void inject(JobComponent component) {
    component.inject(this);
  }

  @Override
  protected void scheduleNext() {
    refreshScheduler.scheduleNext();
  }
}

File: app/src/main/java/org/tasks/jobs/RefreshWork.kt
package org.tasks.jobs

import android.content.Context
import androidx.hilt.work.HiltWorker
import androidx.work.WorkerParameters
import dagger.assisted.Assisted
import dagger.assisted.AssistedInject
import org.tasks.broadcast.RefreshBroadcaster
import org.tasks.analytics.Firebase
import org.tasks.data.dao.TaskDao
import org.tasks.date.DateTimeUtils
import kotlin.math.min

@HiltWorker
class RefreshWork @AssistedInject constructor(
    @Assisted context: Context,
    @Assisted workerParams: WorkerParameters,
    firebase: Firebase,
    private val refreshBroadcaster: RefreshBroadcaster,
    private val workManager: WorkManager,
    private val taskDao: TaskDao,
) : RepeatingWorker(context, workerParams, firebase) {

    override suspend fun run(): Result {
        refreshBroadcaster.broadcastRefresh()
        return Result.success()
    }

    override suspend fun scheduleNext() =
        workManager.scheduleRefresh(min(taskDao.nextRefresh(), DateTimeUtils.midnight()))
}